people/src/hog.cpp

```diff
@@ -171,38 +171,58 @@ pcl::people::HOG::gradHist( float *M, float *O, int h, int w, int bin_size, int
 {
   const int hb=h/bin_size, wb=w/bin_size, h0=hb*bin_size, w0=wb*bin_size, nb=wb*hb;
   const float s=(float)bin_size, sInv=1/s, sInv2=1/s/s;
-  float *H0, *H1, *M0, *M1; int x, y; int *O0, *O1;
+  float *H0, *H1, *M0, *M1; int *O0, *O1;
   O0=(int*)alMalloc(h*sizeof(int),16); M0=(float*) alMalloc(h*sizeof(float),16);
   O1=(int*)alMalloc(h*sizeof(int),16); M1=(float*) alMalloc(h*sizeof(float),16);
 
   // main loop
   float xb = 0;
   float init = 0;
-  for( x=0; x<w0; x++ ) {
+  for(int x=0; x<w0; x++ ) {
     // compute target orientation bins for entire column - very fast
     gradQuantize( O+x*h, M+x*h, O0, O1, M0, M1, n_orients, nb, h0, sInv2 );
 
     if( !soft_bin || bin_size==1 ) {
       // interpolate w.r.t. orientation only, not spatial bin_size
-      H1=H+(x/bin_size)*hb;
-      const auto GH = [&H1, &O0, &O1, &y, &M0, &M1]()
-      {
-          H1[O0[y]]+=M0[y]; H1[O1[y]]+=M1[y]; y++;
+      H1 = H + (x / bin_size) * hb;
+
+      const auto GH = [&H1, &O0, &O1, &M0, &M1](int y) {
+        H1[O0[y]] += M0[y];
+        H1[O1[y]] += M1[y];
       };
-      if( bin_size==1 )      for(y=0; y<h0;) { GH(); H1++; }
-      else if( bin_size==2 ) for(y=0; y<h0;) { GH(); GH(); H1++; }
-      else if( bin_size==3 ) for(y=0; y<h0;) { GH(); GH(); GH(); H1++; }
-      else if( bin_size==4 ) for(y=0; y<h0;) { GH(); GH(); GH(); GH(); H1++; }
-      else for( y=0; y<h0;) { for( int y1=0; y1<bin_size; y1++ ) { GH(); } H1++; }
 
+      for (int y = 0; y < h0; y += bin_size) {
+        // Unrolled loop for small bin_sizes
+        switch (bin_size) {
+        default:
+          for (int inner_loop = 0; inner_loop < bin_size; ++inner_loop) {
+            GH(y);
+          }
+          break;
+        case 4:
+          GH(y);
+          PCL_FALLTHROUGH;
+        case 3:
+          GH(y);
+          PCL_FALLTHROUGH;
+        case 2:
+          GH(y);
+          PCL_FALLTHROUGH;
+        case 1:
+          GH(y);
+          break;
+        }
+        H1++;
+      }
```

Comment:
This logic is different from what was happening before. With each call to `GH(y)`,  `y` was incrementing 1. Right now it is not, which means that if `bin_size` > 1, you're invoking `GH(y)`, letting fallthrough, but `y` is not incremented between each call.

Reply:
```cpp
for (int y = 0; y < h0; ++y)
  GH(y);
H1++;
```
The more I look at it, the more convinced that this would have been enough. :D

Reply:
```cpp
for (int y = 0; y < h0; ++y)
{
  GH(y);
  if (!((y + 1) % (bin_size + 1)))
    H1++;
}
```

Reply:
Nice catch. I missed that

Reply:
So, I'll go ahead and remove the unrolled loop? Or should I keep the unrolled loop?

Reply:
Remove the unrolled loop, it's an eyesore.